### PR TITLE
Support multiple DataTypeAttribute

### PR DESCRIPTION
**Pull request by live1206**
#3356

### CODE REVIEW

File: src/Microsoft.Docs.Template/Models/Attributes.cs [hunk 1]
```diff
@@ -17,10 +17,26 @@ public class DataSchemaAttribute : Attribute { }
...
 
-    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
     public abstract class DataTypeAttribute : Attribute
     {
+        private int _order = -1;
+
         public virtual Type TargetType => typeof(string);
+
+        public int Order
```
Comment (mingwli) [inline src/Microsoft.Docs.Template/Models/Attributes.cs]: is RangeAttribute applicable here?

Comment (live1206) [reply src/Microsoft.Docs.Template/Models/Attributes.cs]: What is RangeAttribute used for?

---
In reply to: [223546032](https://github.com/dotnet/docfx/pull/3479#discussion_r223546032) [](ancestors = 223546032)

Comment (live1206) [reply src/Microsoft.Docs.Template/Models/Attributes.cs]: The idea is from [https://github.com/dotnet/corefx/blob/f5ba053ec0ff8a3a74917f47a56b2927de27473e/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Schema/ColumnAttribute.cs#L47](https://github.com/dotnet/corefx/blob/f5ba053ec0ff8a3a74917f47a56b2927de27473e/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Schema/ColumnAttribute.cs#L47)
We need to sort the attributes based on Order, to apply them accordingly.

---
In reply to: [223546194](https://github.com/dotnet/docfx/pull/3479#discussion_r223546194) [](ancestors = 223546194,223546032)

Comment (yufeih) [reply src/Microsoft.Docs.Template/Models/Attributes.cs]: `Order` should be fixed (get only), markdown is always smaller than xrefproperty

File: src/docfx/build/page/BuildPage.cs [hunk 2]
```diff
@@ -172,8 +172,9 @@ private static (List<Error> errors, Schema schema, PageModel model, JObject meta
 
             return (errors, schema, model, metadata);
 
-            object TransformContent(DataTypeAttribute attribute, object value, string jsonPath)
+            object TransformContent(IEnumerable<DataTypeAttribute> attributes, object value, string jsonPath)
             {
+                var attribute = attributes.OrderBy(attr => attr.Order).First();
                 if (attribute is HrefAttribute)
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: how about using `if (attributes.Any(attr => attr is HrefAttribute)` just like in `BuildXrefMap`? This way you don't even need the `Order` property

Comment (OsmondJiang) [reply src/docfx/build/page/BuildPage.cs]: how about split the XrefProperty and other attributes, add validation to oter attributes

---
In reply to: [223548820](https://github.com/dotnet/docfx/pull/3479#discussion_r223548820) [](ancestors = 223548820)

File: src/docfx/build/xref/XrefMap.cs [hunk 3]
```diff
@@ -56,6 +57,20 @@ public void OutputXrefMap(Context context)
...
+        public static DataTypeAttribute GetNonXrefPropertyAttribute(IEnumerable<DataTypeAttribute> attributes, List<Error> errors, string jsonPath)
+        {
+            try
+            {
+                var otherAttributes = attributes.Where(attr => !(attr is XrefPropertyAttribute));
+                return otherAttributes.SingleOrDefault(attr => !(attr is XrefPropertyAttribute));
+            }
+            catch (InvalidOperationException)
+            {
+                errors.Add(Errors.InvalidMultipleAttributes(jsonPath));
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: we only add error code for end users (writers). for errors introduced by us (developers), just let it throw and make docfx crash.

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: Good to know!

Comment (OsmondJiang) [reply src/docfx/build/xref/XrefMap.cs]: that means the schema is owned by developer?

---
In reply to: [223567812](https://github.com/dotnet/docfx/pull/3479#discussion_r223567812) [](ancestors = 223567812)

Comment (yufeih) [reply src/docfx/build/xref/XrefMap.cs]: yes, there is currently no way for people to extend schemas other than us.

Comment (yufeih) [reply src/docfx/build/xref/XrefMap.cs]: In the future we might add the ability to extend docfx with custom schemas, that is a different scenario and requires extending templates.

**Pull request merged**